### ISSUE TITLE
yum_package version comparison works differently from yum itself

### ISSUE THREAD

**Issue by quulah** [chef/chef#6797]
## Description
yum_package installs older version than yum would install.

We have some internal packages with version strings gathered from issue tracker issue numbers and commit id's. The logic is that packages from the master branch are "newer" than packages from branches. This works when installing with yum by hand, but yum_package installs the branch version incorrectly.

Example of the versions, from master: `1.3+36_ga1cdc0f-199` and from a branch: `1.3~354_g6c80130~ISSUE_117-7`. The + and ~ stuff originates from the deb packages of the same software.

Any way, I tracked the error through the Chef::Provider::Package::Yum stuff, to [rpm_utils](https://github.com/chef/chef/blob/master/lib/chef/provider/package/yum/rpm_utils.rb#L81)

Like the comment says, it's a bit difficult. :)

## Chef Version
13.6.4

## Platform Version
CentOS 7.2

## Replication Case

Minimal example:
```
require 'chef/provider/package/yum/rpm_utils'
Chef::Provider::Package::Yum::RPMUtils.rpmvercmp('1.3+36_ga1cdc0f', '1.3~354_g6c80130~ISSUE_117')
=> -1
```

Compared with the Python library of yum, yum itself and rpmdev-vercmp:
```
>>> import yum
>>> yum.rpmUtils.miscutils.compareVerOnly('1.3+36_ga1cdc0f', '1.3~354_g6c80130~ISSUE_117')
1
```
```
# yum install internal-package
Loaded plugins: fastestmirror
Loading mirror speeds from cached hostfile
...
Resolving Dependencies
--> Running transaction check
---> Package internal-package.noarch 0:1.3~354_g6c80130~ISSUE_117-7.el7 will be updated
---> Package internal-package.noarch 0:1.3+36_ga1cdc0f-199.el7 will be an update
```
```
$ rpmdev-vercmp  1 1.3+36_ga1cdc0f 199 1 1.3~354_g6c80130~ISSUE_117 7
1:1.3+36_ga1cdc0f-199 > 1:1.3~354_g6c80130~ISSUE_117-7
```

Some more things I checked with package objects and RPMDb:
```
master_pkg = Chef::Provider::Package::Yum::RPMDbPackage.new('internal-package', '0', '1.3+36_ga1cdc0f', '199', 'noarch', [], false, false, 'testing-repoid')
branch_pkg = Chef::Provider::Package::Yum::RPMDbPackage.new('internal-package', '0', '1.3~354_g6c80130~ISSUE_117', '7', 'noarch', [], false, false, 'testing-repoid')
master_pkg.compare(branch_pkg)

rpmdb = Chef::Provider::Package::Yum::RPMDb.new()
rpmdb.push(branch_pkg)
rpmdb.push(master_pkg)
rpmdb['internal-package'][0].version
```

**Comment by quulah** [comment]
I can probably go and change the naming scheme of our packages to get this to work.

But I'd expect that Chef's resources treat the versions in the same way as yum.

I'm not sure if there's differences between versions of yum. We're running CentOS 7.

**Comment by thommay** [comment]
@jonlives PTAL

**Comment by lamont-granquist** [comment]
this is almost certainly fixed by #6540 since we no longer use the `RPMUtils.rpmvercmp` API and talk to python instead.

**Comment by lamont-granquist** [comment]
definitely fixed by #6540:

```
[10] pry(main)> Chef::Provider::Package::Yum::PythonHelper.instance.compare_versions('1.3+36_ga1cdc0f', '1.3~354_g6c80130~ISSUE_117')
=> 1
```

**Comment by lamont-granquist** [comment]
think i'm going to close this one, work can be tracked in #6540 which should be fixed in Chef-14.0 (we're pretty much at the point where failure is not an option and it has to ship then)
